diff --git a/node_modules/@definitelytyped/dtslint/dist/index.js b/node_modules/@definitelytyped/dtslint/dist/index.js
index 5850463..3b1531d 100755
--- a/node_modules/@definitelytyped/dtslint/dist/index.js
+++ b/node_modules/@definitelytyped/dtslint/dist/index.js
@@ -121,7 +121,7 @@ async function runTests(dirPath, onlyTestTsNext, expectOnly, tsLocal) {
     const packageName = (0, util_1.packageNameFromPath)(dirPath);
     assertPathIsInDefinitelyTyped(dirPath, dtRoot);
     assertPathIsNotBanned(packageName);
-    assertPackageIsNotDeprecated(packageName, await fs_1.default.promises.readFile((0, path_1.join)(dtRoot, "notNeededPackages.json"), "utf-8"));
+    // assertPackageIsNotDeprecated(packageName, await fs_1.default.promises.readFile((0, path_1.join)(dtRoot, "notNeededPackages.json"), "utf-8"));
     const typesVersions = (0, header_parser_1.getTypesVersions)(dirPath);
     const packageJson = (0, checks_1.checkPackageJson)(dirPath, typesVersions);
     if (Array.isArray(packageJson)) {
