diff --git a/node_modules/@preconstruct/cli/cli/dist/preconstruct-cli-cli.cjs.js b/node_modules/@preconstruct/cli/cli/dist/preconstruct-cli-cli.cjs.js
index b5d4350..42ab524 100644
--- a/node_modules/@preconstruct/cli/cli/dist/preconstruct-cli-cli.cjs.js
+++ b/node_modules/@preconstruct/cli/cli/dist/preconstruct-cli-cli.cjs.js
@@ -2772,7 +2772,8 @@ function getRollupConfigs(pkg) {
           chunkFileNames: `dist/[name]-[hash].${getDistExtensionForConditions(conditions)}`,
           dir: pkg.directory,
           exports: "named",
-          interop
+          // https://github.com/rollup/rollup/issues/5465
+          interop: "esModule"
         }, {
           format: "es",
           entryFileNames: `[name].${getDistExtensionForConditions(conditions.concat("module"))}`,
